client/core/types/chunk.go

```diff
@@ -108,29 +113,49 @@ func (ds *DataSet) Size() uint {
 }
 
 func (ds *DataSet) Count() uint {
+	ds.l.RLock()
+	defer ds.l.RUnlock()
 	return ds.count
 }
 
 func (ds *DataSet) Get(i int) *Chunk {
 	// @TODO (rgeraldes) - add logic to verify if the fragment
 	// exists
+
+	ds.l.RLock()
+	defer ds.l.RUnlock()
 	return ds.data[i]
 }
 
-func (ds *DataSet) Add(chunk *Chunk) {
+func (ds *DataSet) Add(chunk *Chunk) error {
+	if chunk == nil {
```

Comment:
I like this as a safety measure. But would be nice as well to detect when this situation happened and avoid it too.

Reply:
I'm logging the request in https://github.com/kowala-tech/kcoin/pull/613/files#diff-658294a439327f8305b9d63bf426c417R581